src/MessageTrait.php

```diff
@@ -59,59 +66,54 @@ public function getHeaderLine($header)
 
     public function withHeader($header, $value)
     {
-        $new = clone $this;
-        $header = trim($header);
-        $name = strtolower($header);
-
         if (!is_array($value)) {
-            $new->headers[$name] = [trim($value)];
-        } else {
-            $new->headers[$name] = $value;
-            foreach ($new->headers[$name] as &$v) {
-                $v = trim($v);
-            }
+            $value = [$value];
         }
 
-        // Remove the header lines.
-        foreach (array_keys($new->headerLines) as $key) {
-            if (strtolower($key) === $name) {
-                unset($new->headerLines[$key]);
-            }
-        }
+        $normalized = strtolower($header);
 
-        // Add the header line.
-        $new->headerLines[$header] = $new->headers[$name];
+        $new = clone $this;
+        if (isset($new->headerNames[$normalized])) {
+            unset($new->headers[$new->headerNames[$normalized]]);
+        }
+        $new->headerNames[$normalized] = $header;
+        $new->headers[$header] = $value;
 
         return $new;
     }
 
     public function withAddedHeader($header, $value)
     {
-        if (!$this->hasHeader($header)) {
+        $normalized = strtolower($header);
+
+        if (!isset($this->headerNames[$normalized])) {
             return $this->withHeader($header, $value);
         }
 
+        if (!is_array($value)) {
+            $value = [$value];
+        }
+
+        $header = $this->headerNames[$normalized];
+
         $new = clone $this;
-        $new->headers[strtolower($header)][] = $value;
-        $new->headerLines[$header][] = $value;
+        $new->headers[$header] = array_merge($this->headers[$header], $value);
+
         return $new;
     }
 
     public function withoutHeader($header)
     {
-        if (!$this->hasHeader($header)) {
+        $normalized = strtolower($header);
+
+        if (!isset($this->headerNames[$normalized])) {
```

Comment:
I'm not called `hasHeader` here on purpose to prevent two `strtolower` to be executed. This also applies to similar cases